### PR TITLE
Allowlist and denylist when uploading a folder

### PR DESCRIPTION
**Pull request by Wauplin**
Fix https://github.com/huggingface/huggingface_hub/issues/982 and https://github.com/huggingface/huggingface_hub/issues/836.

Implements a `filter_repo_objects` helper to filter objects based on their path and a allowlist/denylist, using [fnmatch](https://docs.python.org/3/library/fnmatch.html). Used in `snapshot_download` and `upload_folder`. Use of `allow_regex`/`ignore_regex` is now deprecated in favor of `allow_patterns`/`ignore_patterns`. Backward compatibility is planned until version 0.12.

**EDIT:** now also implemented in `push_to_hub` mixin and `push_to_hub_keras`.

### CODE REVIEW

File: docs/source/how-to-upstream.mdx [hunk 1]
```diff
@@ -39,24 +39,31 @@ Specify the path of the file to upload, where you want to upload the file to in
...
->>> api.upload_file(path_or_fileobj="/path/to/local/folder/README.md", 
-...                 path_in_repo="README.md", 
-...                 repo_id="username/test-dataset",
-...                 repo_type="dataset",
+>>> api.upload_file(
+...   path_or_fileobj="/path/to/local/folder/README.md",
+...   path_in_repo="README.md",
+...   repo_id="username/test-dataset",
+...   repo_type="dataset",
 ... )
```
Comment (LysandreJik) [inline docs/source/how-to-upstream.mdx:47]: Nitpick, we use 4-space formatting across the repo

File: src/huggingface_hub/utils/_deprecation.py [hunk 2]
```diff
@@ -59,6 +61,9 @@ def _deprecate_arguments(*, version: str, deprecated_args: Set[str]):
             The version when deprecated arguments will result in error.
         deprecated_args (`List[str]`):
             List of the arguments to be deprecated.
+        custom_message (`str`, *optional*):
+            Warning message that is raised. If not passed, a default warning message
```
Comment (LysandreJik) [inline src/huggingface_hub/utils/_deprecation.py:65]: Cool!

**Review by LysandreJik — APPROVED** [review]
Nice, this is a helpful helper! 

Recently we've been moving towards having private module names for modules in `utils` (such as `_deprecations.py`) and importing what we need in the `utils/__init__.py`. Should the same be done with `paths` ?

File: docs/source/how-to-upstream.mdx (hunk 1 above)
Comment (Wauplin) [reply docs/source/how-to-upstream.mdx:47]: Oops yep of course 👍

**Comment by Wauplin** [comment]
Thanks for the review @LysandreJik. I made the small changes in https://github.com/huggingface/huggingface_hub/pull/994/commits/420f6078079bebd4132f26bd471af2e8fbb2794f and will merge once CI is happy.

**Pull request merged**